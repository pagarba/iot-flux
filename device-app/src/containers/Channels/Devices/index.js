import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';

import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';

import { withStyles } from "@material-ui/core/styles/index";

import ChannelDevicesTable from '../../../components/Tables/ChannelDevicesTable';
import CreateConfirmDialog from '../../../components/Dialogs/CreateConfirmDialog';
import { addDeviceToChannel, deleteDeviceFromChannel, getChannel, getChannels, getChannelDevices } from '../../../core/actions/channel'
import { getDevices } from '../../../core/actions/device'

const styles = theme => ({
  root: {
    display: 'block'
  },
  button: {
    margin: theme.spacing.unit,
    marginTop: '2em',
    color: '#fff',
  },
  channelDevicesListPanel: {
    backgroundColor: '#52cee8',
  },
  heading: {
    fontSize: theme.typography.pxToRem(15),
    flexShrink: 0,
    fontWeight: 'bold',
  },
  formControl: {
    margin: theme.spacing.unit,
    minWidth: '120px'
  },
  secondaryHeading: {
    fontSize: theme.typography.pxToRem(15),
    color: theme.palette.text.secondary,
  },
});

export class Devices extends Component {
  state = {
    expanded: 'panel1',
    deviceId: '',
    isCreateConfirmDialogOpen: false,
  };

  componentDidMount() {
    this.props.getDevices();
  }

  handleChange = panel => (event, expanded) => {
    this.setState({
      expanded: expanded ? panel : false,
    });
  };

  handleBlur = event => {
    if (!this.state[event.target.name])
      this.setState({ [event.target.name]: null });
  };

  handleInputChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  handleHideCreateConfirmDialog = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
    })
  }

  navigateToView = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
      expanded: 'panel1'
    });
  }

  navigateToCreate = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
      expanded: 'panel2'
    });
  }

  addDeviceToChannel = () => {
    const { deviceId } = this.state;
    const { id: channelId } = this.props.channel;

    if (deviceId) {
      this.props.addDeviceToChannel(
        channelId,
        deviceId
      ).then(() => {
        this.props.getChannel(channelId);
        this.props.getChannelDevices(channelId);
        this.setState({
          deviceId: '',
          isCreateConfirmDialogOpen: true,
        })
      });
    }
  };

  deleteDeviceFromChannel = (deviceId) => {
    const { id } = this.props.channel;

    this.props.deleteDeviceFromChannel(id, deviceId)
      .then(() => {
        this.props.getChannel(id);
      })
      .catch(() => {
        alert('Error while disconnecting device');
      });
  }

  render() {
    const { classes, history, channelDevices } = this.props;
    const { expanded } = this.state;
    const connectedDevices = (channelDevices || {}).things || [];

    let devices = this.props.devices ? this.props.devices.things.filter(dev1 => 0 === connectedDevices.filter(dev2 => dev1.id === dev2.id).length) : [];

    return (
      <div className={classes.root}>
        <ExpansionPanel
          className={classes.channelDevicesListPanel}
          expanded={expanded === 'panel1'}
          onChange={this.handleChange('panel1')}
        >
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon color="primary"/>}>
            <Typography color="primary" className={classes.heading}>CONNECTED DEVICES</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails>
            <ChannelDevicesTable
              data={connectedDevices}
              history={history}
              onDeleteDeviceFromChannel={this.deleteDeviceFromChannel}
            />
          </ExpansionPanelDetails>
        </ExpansionPanel>
        <ExpansionPanel expanded={expanded === 'panel2'} onChange={this.handleChange('panel2')}>
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
            <Typography className={classes.heading}>CONNECT A NEW DEVICE</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails className={classes.root}>
            <div>
              <FormControl className={classes.formControl}>
                <InputLabel htmlFor="deviceId">Device</InputLabel>
                <Select
                  value={this.state.deviceId}
                  onChange={this.handleInputChange}
                  inputProps={{
                    name: 'deviceId',
                    id: 'deviceId',
                  }}
                >
                  {
                    devices.map((device) => (
                      <MenuItem value={device.id}>{device.name}</MenuItem>
                    ))
                  }
                </Select>
              </FormControl>
            </div>
            <Button
              className={classes.button}
              color="secondary"
              onClick={this.addDeviceToChannel}
              variant="contained"
              disabled={this.props.isAddingDeviceToChannel}
            >
              Connect Device
            </Button>
          </ExpansionPanelDetails>
        </ExpansionPanel>
        <CreateConfirmDialog
          open={this.state.isCreateConfirmDialogOpen}
          onClickDismiss={this.handleHideCreateConfirmDialog}
          onClickView={this.navigateToView}
          onClickCreate={this.navigateToCreate}
          contentText="Device was added to channel"
          titleText="Confirm"
          viewButtonText="View Devices"
          createButtonText="Connect Device"
        />
      </div>
    )
  }
}

Devices.propTypes = {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
};

function mapStateToProps(state) {
  return {
    devices: state.rootReducer.device.devices,
    channelDevices: state.rootReducer.channel.channelDevices,
    isAddingDeviceToChannel: state.rootReducer.channel.isAddingDeviceToChannel,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    addDeviceToChannel: (channelId, deviceId) => dispatch(addDeviceToChannel(channelId, deviceId)),
    deleteDeviceFromChannel: (channelId, deviceId) => dispatch(deleteDeviceFromChannel(channelId, deviceId)),
    getChannels: () => dispatch(getChannels()),
    getChannelDevices: (deviceId) => dispatch(getChannelDevices(deviceId)),
    getDevices: () => dispatch(getDevices()),
    getChannel: (id) => dispatch(getChannel(id)),
  }
}

const WithStyles =  withStyles(styles, { withTheme: true })(Devices);
export default connect(mapStateToProps, mapDispatchToProps)(WithStyles);