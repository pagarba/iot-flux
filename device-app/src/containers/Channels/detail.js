import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';

import { withStyles } from "@material-ui/core/styles/index";
import {getChannel, editChannel, getChannelDevices} from '../../core/actions/channel'
import ChannelDevices from './Devices';

const styles = theme => ({
  root: {
    display: 'block',
  },
  button: {
    margin: theme.spacing.unit,
    marginTop: '2em',
    color: '#fff',
  },
  heading: {
    fontSize: theme.typography.pxToRem(15),
    flexBasis: '33.33%',
    flexShrink: 0,
    fontWeight: 'bold',
  },
  formControl: {
    margin: theme.spacing.unit,
  },
  secondaryHeading: {
    fontSize: theme.typography.pxToRem(15),
    color: theme.palette.text.secondary,
  },
});

export class ChannelDetail extends Component {
  constructor () {
    super();

    this.state = {
      isEditMode: false,
      name: '',
    };
  }

  componentWillMount() {
    const channelId = this.props.match.params.channelId;

    this.props.getChannel(channelId);
    this.props.getChannelDevices(channelId);
  }

  componentWillReceiveProps(props) {
    this.setState({
      name: props.channel.name,
    })
  }

  editChannel = () => {
    const { name } = this.state;
    const channelId = this.props.match.params.channelId;

    this.props.editChannel(channelId, {
      name
    }).then(() => {
      this.props.getChannel(channelId).then(() => {
        this.setState({
          isEditMode: false,
        })
      });
    });
  };

  handleInputChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  render() {
    const { classes } = this.props;

    const channel = this.props.channel || {};

    return (
      <div className={classes.root}>
        <ChannelDevices
          channel={channel}
        />
      </div>
    )
  }
}

ChannelDetail.propTypes = {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
};

function mapStateToProps(state) {
  return {
    channel: state.rootReducer.channel.channel,
    isEditingChannel: state.rootReducer.channel.isEditingChannel,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    getChannel: (id) => dispatch(getChannel(id)),
    editChannel: (id, data) => dispatch(editChannel(id, data)),
    getChannelDevices: (deviceId) => dispatch(getChannelDevices(deviceId)),
  }
}

const WithStyles =  withStyles(styles, { withTheme: true })(ChannelDetail);
export default connect(mapStateToProps, mapDispatchToProps)(WithStyles);
