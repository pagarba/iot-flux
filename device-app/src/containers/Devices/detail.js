import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';

import FormControl from '@material-ui/core/FormControl';
import Input from '@material-ui/core/Input';

import { withStyles } from "@material-ui/core/styles/index";
import { getDevice, editDevice } from '../../core/actions/device';

const styles = theme => ({
  root: {
    display: 'block'
  },
  button: {
    margin: theme.spacing.unit,
    marginTop: '2em',
    color: '#fff',
  },
  heading: {
    fontSize: theme.typography.pxToRem(15),
    flexBasis: '33.33%',
    flexShrink: 0,
    fontWeight: 'bold',
  },
  formControl: {
    margin: theme.spacing.unit,
  },
  secondaryHeading: {
    fontSize: theme.typography.pxToRem(15),
    color: theme.palette.text.secondary,
  },
});

export class DeviceDetail extends Component {
  constructor () {
    super();

    this.state = {
      isEditMode: false,
      key: '',
      name: '',
      type: '',
    };
  }

  componentWillMount() {
    const deviceId = this.props.match.params.deviceId;
    this.props.getDevice(deviceId);
  }

  componentWillReceiveProps(props) {
    this.setState({
      key: props.device.key,
      name: props.device.name,
      type: props.device.type,
    })
  }

  editDevice = () => {
    const { name, type } = this.state;
    const deviceId = this.props.match.params.deviceId;

    this.props.editDevice(deviceId, {
      name,
      type
    }).then(() => {
      this.props.getDevice(deviceId).then(() => {
        this.setState({
          isEditMode: false,
        })
      });
    });
  }

  handleInputChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  render() {
    const { classes } = this.props;
    const { isEditMode } = this.state;

    const device = this.props.device || {};

    return (
      <div className={classes.root}>
        <table>
          <tbody>
            <tr>
              <td><b>Device ID: </b></td>
              <td>
                {device.id}
              </td>
            </tr>
            <tr>
              <td><b>Device Key: </b></td>
              <td>{ device.key }</td>
            </tr>
            <tr>
              <td><b>Device Type: </b></td>
              <td>{device.type}</td>
            </tr>
            <tr>
              <td><b>Device Name: </b></td>
              <td>
                {
                  isEditMode ? (
                    <FormControl className={classes.formControl} error={this.state.name === null} fullWidth aria-describedby="component-error-text">
                      <Input name="name" value={this.state.name} onChange={this.handleInputChange} />
                    </FormControl>
                  ) : device.name
                }
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    )
  }
}

DeviceDetail.propTypes = {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
};

function mapStateToProps(state) {
  return {
    device: state.rootReducer.device.device,
    isEditingDevice: state.rootReducer.device.isEditingDevice,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    getDevice: (id) => dispatch(getDevice(id)),
    editDevice: (id, data) => dispatch(editDevice(id, data)),
  }
}

const WithStyles =  withStyles(styles, { withTheme: true })(DeviceDetail);
export default connect(mapStateToProps, mapDispatchToProps)(WithStyles);
