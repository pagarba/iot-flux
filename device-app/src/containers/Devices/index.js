import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';

import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';

import { withStyles } from "@material-ui/core/styles/index";

import DevicesTable from '../../components/Tables/DevicesTable';
import CreateConfirmDialog from '../../components/Dialogs/CreateConfirmDialog';
import { createDevice, getDevices, deleteDevice, editDevice } from '../../core/actions/device'
import qs from 'stringquery'

const styles = theme => ({
  root: {
    display: 'block'
  },
  button: {
    margin: theme.spacing.unit,
    marginTop: '2em',
    color: '#fff',
  },
  devicesListPanel: {
    backgroundColor: '#52cee8',
  },
  heading: {
    fontSize: theme.typography.pxToRem(15),
    flexShrink: 0,
    fontWeight: 'bold',
  },
  formControl: {
    margin: theme.spacing.unit,
  },
  secondaryHeading: {
    fontSize: theme.typography.pxToRem(15),
    color: theme.palette.text.secondary,
  },
});

export class Devices extends Component {
  state = {
    expanded: null,
    deviceType: '',
    deviceName: '',
    deviceKey: '',
    isCreateConfirmDialogOpen: false,
  };

  componentWillMount() {
    this.props.getDevices();
  }

  componentDidMount() {
    const obj = qs(this.props.location.search);
    this.setState({
      expanded: obj.mode === 'create' ? 'panel2' : 'panel1'
    })
  }

  handleChange = panel => (event, expanded) => {
    this.setState({
      expanded: expanded ? panel : false,
    });
  };

  handleBlur = event => {
    if (!this.state[event.target.name])
      this.setState({ [event.target.name]: null });
  };

  handleHideCreateConfirmDialog = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
    })
  }

  handleInputChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  navigateToView = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
      expanded: 'panel1'
    });
  }

  navigateToCreate = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
      expanded: 'panel2'
    });
  }

  addDevice = () => {
    const { deviceName, deviceType } = this.state;

    if (deviceName && deviceType) {
      this.props.createDevice({
        type: deviceType,
        name: deviceName,
      }).then(() => {
        this.props.getDevices();
        this.setState({
          deviceName: '',
          deviceType: '',
          isCreateConfirmDialogOpen: true,
        })
      });
    }
  };

  deleteDevice = (id) => {
    this.props.deleteDevice(id)
      .then(() => {
        this.props.getDevices();
      })
      .catch(() => {
        alert('Error while deleting device');
      });
  }

  render() {
    const { classes, history } = this.props;
    const { expanded } = this.state;

    return (
      <div className={classes.root}>
        <ExpansionPanel
          className={classes.devicesListPanel}
          expanded={expanded === 'panel1'}
          onChange={this.handleChange('panel1')}
        >
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon color="primary"/>}>
            <Typography color="primary" className={classes.heading}>MY DEVICES</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails>
            <DevicesTable
              history={history}
              data={this.props.devices ? this.props.devices.things : []}
              editDevice={this.props.editDevice}
              getDevices={this.props.getDevices}
              onDeleteDevice={this.deleteDevice}
            />
          </ExpansionPanelDetails>
        </ExpansionPanel>
        <ExpansionPanel expanded={expanded === 'panel2'} onChange={this.handleChange('panel2')}>
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
            <Typography className={classes.heading}>ADD A NEW DEVICE</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails className={classes.root}>
            <FormControl className={classes.formControl} error={this.state.deviceName === null} fullWidth aria-describedby="component-error-text">
              <InputLabel htmlFor="component-error">Device Name*</InputLabel>
              <Input name="deviceName" value={this.state.deviceName} onChange={this.handleInputChange} onBlur={this.handleBlur} />
              <FormHelperText id="component-error-text">This field is required</FormHelperText>
            </FormControl>

            <FormControl className={classes.formControl} fullWidth>
              <InputLabel htmlFor="deviceType">Device Type</InputLabel>
              <Select
                value={this.state.deviceType}
                onChange={this.handleInputChange}
                inputProps={{
                  name: 'deviceType',
                  id: 'deviceType',
                }}
              >
                <MenuItem value="device">Device</MenuItem>
                <MenuItem value="app">App</MenuItem>
              </Select>
            </FormControl>

            <Button
              className={classes.button}
              color="secondary"
              onClick={this.addDevice}
              variant="contained"
              disabled={this.props.isCreatingDevice}
            >
              Add Device
            </Button>
          </ExpansionPanelDetails>
        </ExpansionPanel>
        <CreateConfirmDialog
          open={this.state.isCreateConfirmDialogOpen}
          onClickDismiss={this.handleHideCreateConfirmDialog}
          onClickView={this.navigateToView}
          onClickCreate={this.navigateToCreate}
          contentText="A new device has been created"
          titleText="Confirm"
          viewButtonText="View Devices"
          createButtonText="Create Device"
        />
      </div>
    )
  }
}

Devices.propTypes = {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
};

function mapStateToProps(state) {
  return {
    devices: state.rootReducer.device.devices,
    isCreatingDevice: state.rootReducer.device.isCreatingDevice,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    getDevices: () => dispatch(getDevices()),
    createDevice: data => dispatch(createDevice(data)),
    deleteDevice: id => dispatch(deleteDevice(id)),
    editDevice: (id, data) => dispatch(editDevice(id, data)),
  }
}

const WithStyles =  withStyles(styles, { withTheme: true })(Devices);
export default connect(mapStateToProps, mapDispatchToProps)(WithStyles);