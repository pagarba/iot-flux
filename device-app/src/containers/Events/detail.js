import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux'

import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import InputAdornment from '@material-ui/core/InputAdornment';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';

import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import SearchIcon from '@material-ui/icons/Search';

import { withStyles } from "@material-ui/core/styles/index";

import { getChannel, getChannels, getChannelDevices } from '../../core/actions/channel';
import { createEvent, getEvents, emptyEvents } from '../../core/actions/event';
import { getDevices } from '../../core/actions/device';

import EventsTable from '../../components/Tables/EventsTable';
import CreateConfirmDialog from '../../components/Dialogs/CreateConfirmDialog';
import moment from "moment/moment";

const styles = theme => ({
  root: {
    display: 'block'
  },
  button: {
    margin: theme.spacing.unit,
    marginTop: '2em',
    color: '#fff',
  },
  channelFilter: {
    backgroundColor: '#fff',
    width: '100%',
    display: 'flex',
    alignItems: 'flex-end',
    borderTopLeftRadius: '4px',
    borderTopRightRadius: '4px',
  },
  eventsListPanel: {
    backgroundColor: '#52cee8',
  },
  eventsListPanelDetails: {
    display: 'flex',
    flexDirection: 'column',
  },
  heading: {
    fontSize: theme.typography.pxToRem(15),
    flexShrink: 0,
    fontWeight: 'bold',
  },
  formControl: {
    minWidth: '150px',
    margin: '0 1rem',
  },
  secondaryHeading: {
    fontSize: theme.typography.pxToRem(15),
    color: theme.palette.text.secondary,
  },
});

export class EventsDetail extends Component {
  state = {
    channelId: '',
    expanded: 'panel1',
    eventName: '',
    deviceKey: '',
    data: '',
    isCreateConfirmDialogOpen: false,
    searchKeyword: '',
  };

  async componentDidMount() {

    /**
     * Latest way to get channel devices. (With dropdown and with updated mainflux API)
     */
    this.props.getDevices();
    await this.props.getChannels();

    let { channels } = this.props;
    channels = channels.channels;

    if (channels.length) {
      await this.props.getChannelDevices(channels[0].id);

      this.setState({
        channelId: channels[0].id,
      });

      const { channelDevices } = this.props;
      if (channelDevices && channelDevices.things && channelDevices.things.length) {
        const deviceKey = channelDevices.things[0].key;

        this.props.getEvents(channels[0].id, deviceKey);
      } else {
        this.props.getEvents(channels[0].id, 'random key');
      }
    } else {
      this.props.emptyEvents();
    }
  }

  addEvent = () => {
    const { channelId, deviceKey, data } = this.state;

    if (data && deviceKey) {
      this.props.createEvent(channelId, deviceKey, data)
        .then(() => {
          this.props.getEvents(channelId, deviceKey);
          this.setState({
            deviceKey: '',
            data: '',
            isCreateConfirmDialogOpen: true,
          })
        });
    }
  };

  handleChange = panel => (event, expanded) => {
    this.setState({
      expanded: expanded ? panel : false,
    });
  };

  handleBlur = event => {
    if (!this.state[event.target.name])
      this.setState({ [event.target.name]: null });
  };

  handleHideCreateConfirmDialog = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
    })
  };

  handleInputChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  handleChannelChange = async (event) => {
    this.setState({
      channelId: event.target.value,
    });

    this.props.getChannel(event.target.value);
    await this.props.getChannelDevices(event.target.value);

    const { channelDevices } = this.props;

    if (channelDevices && channelDevices.things && channelDevices.things.length) {
      const deviceKey = channelDevices.things[0].key;
      this.props.getEvents(event.target.value, deviceKey);
    } else {
      this.props.getEvents(event.target.value, 'random key');
    }
  };

  navigateToView = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
      expanded: 'panel1'
    });
  };

  navigateToCreate = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
      expanded: 'panel2'
    });
  };

  filteredMessages = () => {
    const { events, devices } = this.props;
    const { searchKeyword } = this.state;

    return ((events || {}).messages || []).filter((e) => {
      const deviceName = devices && devices.things ? (devices.things.find((device) => device.id === e.publisher) || {}).name : '';
      return e.name.indexOf(searchKeyword) !== -1 || e.protocol.indexOf(searchKeyword) !== -1
        || e.unit.indexOf(searchKeyword) !== -1 || e.value.toString().indexOf(searchKeyword) !== -1
        || moment(e.time).format('YYYY-MM-DD HH:mm:ss').indexOf(searchKeyword) !== -1
        || deviceName.indexOf(searchKeyword) !== -1;
    });
  };

  render() {
    const { classes, history, channels, devices, channelDevices } = this.props;
    const { expanded } = this.state;

    return (
      <div className={classes.root}>
        <ExpansionPanel
          className={classes.eventsListPanel}
          expanded={expanded === 'panel1'}
          onChange={this.handleChange('panel1')}
        >
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon color="primary" />}>
            <Typography className={classes.heading} color="primary">MY EVENTS</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails className={classes.eventsListPanelDetails}>
            <div className={classes.channelFilter}>
              <FormControl className={classes.formControl}>
                <InputLabel htmlFor="deviceKey">Select Channel</InputLabel>
                <Select
                  value={this.state.channelId}
                  onChange={this.handleChannelChange}
                  inputProps={{
                    name: 'channelId',
                    id: 'channelId',
                  }}
                >
                  {
                    ((channels || {}).channels || []).map((channel) => (
                      <MenuItem value={channel.id}>{channel.name}</MenuItem>
                    ))
                  }
                </Select>
              </FormControl>
              <FormControl className={classes.formControl}>
                <Input
                  id="searchKeyword"
                  name="searchKeyword"
                  onChange={this.handleInputChange}
                  startAdornment={
                    <InputAdornment position="start">
                      <SearchIcon />
                    </InputAdornment>
                  }
                />
              </FormControl>
            </div>
            <EventsTable
              history={history}
              data={this.filteredMessages()}
              devices={devices}
            />
          </ExpansionPanelDetails>
        </ExpansionPanel>
        <ExpansionPanel expanded={expanded === 'panel2'} onChange={this.handleChange('panel2')}>
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
            <Typography className={classes.heading}>ADD A NEW EVENT</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails className={classes.root}>
            <FormControl className={classes.formControl}>
              <InputLabel htmlFor="deviceKey">Device</InputLabel>
              <Select
                value={this.state.deviceKey}
                onChange={this.handleInputChange}
                inputProps={{
                  name: 'deviceKey',
                  id: 'deviceKey',
                }}
              >
                {
                  channelDevices && channelDevices.things && channelDevices.things.map((device) => (
                    <MenuItem value={device.key}>{device.name}</MenuItem>
                  ))
                }
              </Select>
            </FormControl>

            <FormControl className={classes.formControl} error={this.state.data === null} fullWidth aria-describedby="component-error-text">
              <InputLabel htmlFor="component-error">Data</InputLabel>
              <Input name="data" value={this.state.data} onChange={this.handleInputChange} onBlur={this.handleBlur} />
              <FormHelperText id="component-error-text">This field is required</FormHelperText>
            </FormControl>

            <Button
              className={classes.button}
              color="secondary"
              onClick={this.addEvent}
              variant="contained"
              disabled={this.props.isCreatingEvent}
            >
              Add Event
            </Button>
          </ExpansionPanelDetails>
        </ExpansionPanel>
        <CreateConfirmDialog
          open={this.state.isCreateConfirmDialogOpen}
          onClickDismiss={this.handleHideCreateConfirmDialog}
          onClickView={this.navigateToView}
          onClickCreate={this.navigateToCreate}
          contentText="A new event has been created"
          titleText="Confirm"
          viewButtonText="View Events"
          createButtonText="Create Event"
        />
      </div>
    )
  }
}

EventsDetail.propTypes = {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
};

function mapStateToProps(state) {
  return {
    channel: state.rootReducer.channel.channel,
    channels: state.rootReducer.channel.channels,
    channelDevices: state.rootReducer.channel.channelDevices,
    events: state.rootReducer.event.events,
    devices: state.rootReducer.device.devices,
    isCreatingEvent: state.rootReducer.event.isCreatingEvent,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    createEvent: (channelId, deviceKey, data) => dispatch(createEvent(channelId, deviceKey, data)),
    getChannel: (id) => dispatch(getChannel(id)),
    getChannels: () => dispatch(getChannels()),
    getChannelDevices: (id) => dispatch(getChannelDevices(id)),
    getDevices: () => dispatch(getDevices()),
    getEvents: (channelId, deviceKey) => dispatch(getEvents(channelId, deviceKey)),
    emptyEvents: () => dispatch(emptyEvents()),
  }
}

const WithStyles =  withStyles(styles, { withTheme: true })(EventsDetail);
export default connect(mapStateToProps, mapDispatchToProps)(WithStyles);
