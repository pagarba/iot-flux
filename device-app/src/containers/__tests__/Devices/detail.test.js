import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import { DeviceDetail } from '../../Devices/detail';

describe('<DeviceDetail />', () => {
  let shallow;

  beforeAll(() => {
    shallow = createShallow();
  });

  it('should render components', () => {
    const wrapper = shallow(
      <DeviceDetail
        classes={{}}
        match={{params: {}}}
        getDevice={jest.fn()}
        theme={{}}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});