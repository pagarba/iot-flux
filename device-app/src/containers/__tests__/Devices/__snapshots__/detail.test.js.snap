// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeviceDetail /> should render components 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td>
          <b>
            Device ID: 
          </b>
        </td>
        <td />
      </tr>
      <tr>
        <td>
          <b>
            Device Key: 
          </b>
        </td>
        <td />
      </tr>
      <tr>
        <td>
          <b>
            Device Type: 
          </b>
        </td>
        <td />
      </tr>
      <tr>
        <td>
          <b>
            Device Name: 
          </b>
        </td>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;
